From: Jiri Slaby <jslaby@suse.cz>
Date: Sun, 26 Aug 2018 14:44:15 +0200
Subject: Revert "netfilter: ipv6: nf_defrag: reduce struct net memory waste"
Patch-mainline: never, kabi
References: kabi

This reverts commit 8747d9e7d45420350524dc8a8343837b8ac92776, upstream
commit 9ce7bc036ae4cfe3393232c86e9e1fea2153c237. It shuffles with
members of struct net and struct netns_nf_frag only for optimizations.
This broke kABI, so revert this.

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 include/net/net_namespace.h             |    1 -
 include/net/netns/ipv6.h                |    1 +
 net/ipv6/netfilter/nf_conntrack_reasm.c |    6 +++---
 3 files changed, 4 insertions(+), 4 deletions(-)

--- a/include/net/net_namespace.h
+++ b/include/net/net_namespace.h
@@ -115,7 +115,6 @@ struct net {
 #endif
 #if IS_ENABLED(CONFIG_NF_DEFRAG_IPV6)
 	struct netns_nf_frag	nf_frag;
-	struct ctl_table_header *nf_frag_frags_hdr;
 #endif
 	struct sock		*nfnl;
 	struct sock		*nfnl_stash;
--- a/include/net/netns/ipv6.h
+++ b/include/net/netns/ipv6.h
@@ -86,6 +86,7 @@ struct netns_ipv6 {
 
 #if IS_ENABLED(CONFIG_NF_DEFRAG_IPV6)
 struct netns_nf_frag {
+	struct netns_sysctl_ipv6 sysctl;
 	struct netns_frags	frags;
 };
 #endif
--- a/net/ipv6/netfilter/nf_conntrack_reasm.c
+++ b/net/ipv6/netfilter/nf_conntrack_reasm.c
@@ -118,7 +118,7 @@ static int nf_ct_frag6_sysctl_register(s
 	if (hdr == NULL)
 		goto err_reg;
 
-	net->nf_frag_frags_hdr = hdr;
+	net->nf_frag.sysctl.frags_hdr = hdr;
 	return 0;
 
 err_reg:
@@ -132,8 +132,8 @@ static void __net_exit nf_ct_frags6_sysc
 {
 	struct ctl_table *table;
 
-	table = net->nf_frag_frags_hdr->ctl_table_arg;
-	unregister_net_sysctl_table(net->nf_frag_frags_hdr);
+	table = net->nf_frag.sysctl.frags_hdr->ctl_table_arg;
+	unregister_net_sysctl_table(net->nf_frag.sysctl.frags_hdr);
 	if (!net_eq(net, &init_net))
 		kfree(table);
 }
